File: long_questions/java/core-java/coding-practice/100-java-programs/Java_Collection/CollectionDemo.java
import java.util.*;

/**
 * Demonstrates Collection Interface common methods.
 * 
 * Methods covered:
 * - add(E e), addAll(Collection c)
 * - remove(Object o), removeAll(Collection c), retainAll(Collection c)
 * - clear(), size(), isEmpty()
 * - contains(Object o), containsAll(Collection c)
 * - toArray()
 * - stream(), parallelStream()
 */
public class CollectionDemo {

    public static void main(String[] args) {
        // Using ArrayList as the Collection implementation
        Collection<String> collection = new ArrayList<>();

        System.out.println("--- Adding Elements ---");
        // add(E e)
        collection.add("Apple");
        collection.add("Banana");
        collection.add("Cherry");
        System.out.println("Collection after add: " + collection);

        // addAll(Collection c)
        Collection<String> moreFruits = new ArrayList<>();
        moreFruits.add("Date");
        moreFruits.add("Elderberry");
        collection.addAll(moreFruits);
        System.out.println("Collection after addAll: " + collection);

        System.out.println("\n--- Checking Status ---");
        // size()
        System.out.println("Size: " + collection.size());
        // isEmpty()
        System.out.println("Is Empty: " + collection.isEmpty());
        // contains(Object o)
        System.out.println("Contains 'Banana': " + collection.contains("Banana"));
        // containsAll(Collection c)
        System.out.println("Contains all 'moreFruits': " + collection.containsAll(moreFruits));

        System.out.println("\n--- Removing Elements ---");
        // remove(Object o)
        collection.remove("Apple");
        System.out.println("Removed 'Apple': " + collection);

        // removeAll(Collection c)
        Collection<String> toRemove = new ArrayList<>();
        toRemove.add("Date");
        collection.removeAll(toRemove);
        System.out.println("Removed 'Date' (removeAll): " + collection);

        // retainAll(Collection c) (Intersection)
        Collection<String> toRetain = new ArrayList<>();
        toRetain.add("Banana");
        toRetain.add("Cherry");
        collection.retainAll(toRetain);
        System.out.println("Retained only 'Banana' and 'Cherry': " + collection);

        System.out.println("\n--- Conversion ---");
        // toArray()
        Object[] array = collection.toArray();
        System.out.println("Converted to Array: " + Arrays.toString(array));

        System.out.println("\n--- Streams ---");
        // stream()
        System.out.println("Stream processing:");
        collection.stream()
            .filter(s -> s.startsWith("B"))
            .forEach(System.out::println);

        System.out.println("\n--- Clearing ---");
        // clear()
        collection.clear();
        System.out.println("Collection after clear: " + collection);
        System.out.println("Is Empty now: " + collection.isEmpty());
    }
}

File: long_questions/golang/core-go/coding-practice/100-go-programs/Go_Collection/CollectionDemo.go
package main

import "fmt"

/**
 * Demonstrates Go slice operations equivalent to Java Collection Interface.
 * 
 * Methods covered:
 * - append(), copy()
 * - manual remove, filter
 * - len(), cap()
 * - contains, containsAll
 * - to array conversion
 */

func main() {
    // Using slice as the Collection implementation
    collection := []string{}

    fmt.Println("--- Adding Elements ---")
    // append equivalent to add(E e)
    collection = append(collection, "Apple")
    collection = append(collection, "Banana")
    collection = append(collection, "Cherry")
    fmt.Printf("Collection after add: %v\n", collection)

    // append equivalent to addAll(Collection c)
    moreFruits := []string{"Date", "Elderberry"}
    collection = append(collection, moreFruits...)
    fmt.Printf("Collection after addAll: %v\n", collection)

    fmt.Println("\n--- Checking Status ---")
    // len() equivalent to size()
    fmt.Printf("Size: %d\n", len(collection))
    // isEmpty equivalent
    fmt.Printf("Is Empty: %t\n", len(collection) == 0)
    
    // contains equivalent
    fmt.Printf("Contains 'Banana': %t\n", contains(collection, "Banana"))
    fmt.Printf("Contains all 'moreFruits': %t\n", containsAll(collection, moreFruits))

    fmt.Println("\n--- Removing Elements ---")
    // remove equivalent (manual implementation)
    collection = remove(collection, "Apple")
    fmt.Printf("Removed 'Apple': %v\n", collection)

    // removeAll equivalent
    toRemove := []string{"Banana", "Date"}
    collection = removeAll(collection, toRemove)
    fmt.Printf("Removed multiple: %v\n", collection)

    fmt.Println("\n--- Other Operations ---")
    // clear equivalent
    collection = collection[:0] // Clear slice
    fmt.Printf("After clear: %v\n", collection)
    
    // toArray equivalent
    newCollection := []string{"X", "Y", "Z"}
    array := toArray(newCollection)
    fmt.Printf("To array: %v\n", array)
}

// Helper functions
func contains(slice []string, item string) bool {
    for _, s := range slice {
        if s == item {
            return true
        }
    }
    return false
}

func containsAll(slice, subSlice []string) bool {
    for _, item := range subSlice {
        if !contains(slice, item) {
            return false
        }
    }
    return true
}

func remove(slice []string, item string) []string {
    for i, s := range slice {
        if s == item {
            return append(slice[:i], slice[i+1:]...)
        }
    }
    return slice
}

func removeAll(slice, toRemove []string) []string {
    result := []string{}
    toRemoveSet := make(map[string]bool)
    for _, item := range toRemove {
        toRemoveSet[item] = true
    }
    
    for _, item := range slice {
        if !toRemoveSet[item] {
            result = append(result, item)
        }
    }
    return result
}

func toArray(slice []string) []string {
    // In Go, slices are already backed by arrays
    // This is just a copy demonstration
    result := make([]string, len(slice))
    copy(result, slice)
    return result
}
